gamemaster/local.go

```diff
@@ -19,37 +21,134 @@ type update struct {
 type localEngine struct {
 	state    game.State
 	updateCh chan update
+	gameOver bool
 }
 
-// TODO: single instance for all players => use factory pattern?
+var (
+	singleLocalEngine *localEngine
+	once              sync.Once
+)
+
 func NewLocalEngine() *localEngine {
 	return &localEngine{}
 }
 
-// TODO: remote engine should also satisfy this interface
+// @Zhou to use this, replace any calls like: engine := NewLocalEngine() with engine := gamemaster.GetLocalEngine()
+func GetLocalEngine() *localEngine {
+	once.Do(func() {
+		singleLocalEngine = &localEngine{}
+	})
+	return singleLocalEngine
+}
+
 func (e *localEngine) Init() (game.State, UpdateGetter) {
```

Comment:
- `LocalEngine() Engine` 1) as a factory method that returns an Engine singleton 2) max called twice to initialize a game for 2 players 3) adds a player each time it's called 4) initialize the game state when it's first called
- `engine.GetUpdate() (Move, State)`